zkevm-circuits/src/state_circuit/binary_number.rs

```diff
@@ -0,0 +1,172 @@
+        move |meta: &mut VirtualCells<'_, F>| {
+            let bits = self.bits.map(|bit| meta.query_advice(bit, rotation));
+            bits.iter()
+                .fold(0.expr(), |result, bit| bit.clone() + result * 2.expr())
+        }
+    }
+
+    pub fn value_equals<F: Field, S: AsBits<N>>(
+        &self,
+        value: &S,
+        rotation: Rotation,
+    ) -> impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F> + '_ {
+        let bits = value.as_bits();
+        move |meta: &mut VirtualCells<'_, F>| {
+            Self::value_equals_expr(
+                &bits,
+                &self.bits.map(|bit| meta.query_advice(bit, rotation)),
+            )
+        }
+    }
+
+    pub fn value_equals_expr<F: Field>(
+        value: &[bool],
+        expressions: &[Expression<F>],
+    ) -> Expression<F> {
+        and::expr(value.iter().zip(expressions).map(|(&bit, expression)| {
+            if bit {
+                expression.clone()
+            } else {
+                not::expr(expression.clone())
+            }
+        }))
+    }
+}
+
+pub struct Chip<F: Field, T, const N: usize>
+where
+    T: AsBits<N>,
+{
+    config: Config<T, N>,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field, T: IntoEnumIterator, const N: usize> Chip<F, T, N>
+where
+    T: AsBits<N>,
+{
+    pub fn construct(config: Config<T, N>) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn configure(meta: &mut ConstraintSystem<F>, selector: Column<Fixed>) -> Config<T, N> {
+        let bits = [0; N].map(|_| meta.advice_column());
+        bits.map(|bit| {
+            meta.create_gate("bit column is 0 or 1", |meta| {
+                let selector = meta.query_fixed(selector, Rotation::cur());
+                let bit = meta.query_advice(bit, Rotation::cur());
+                vec![selector * bit.clone() * (1.expr() - bit)]
+            })
+        });
+
+        let config = Config {
+            bits,
+            _marker: PhantomData,
+        };
+
+        // Disallow bit patterns (if any) that don't correspond to a variant of T.
+        let valid_values: BTreeSet<usize> = T::iter().map(|t| from_bits(&t.as_bits())).collect();
+        let mut invalid_values = (0..1 << N).filter(|i| !valid_values.contains(i)).peekable();
+        if invalid_values.peek().is_some() {
+            meta.create_gate("binary number value in range", |meta| {
+                let selector = meta.query_fixed(selector, Rotation::cur());
+                invalid_values
+                    .map(|i| {
+                        let value_equals_i = config.value_equals(&i, Rotation::cur());
+                        selector.clone() * value_equals_i(meta)
+                    })
+                    .collect::<Vec<_>>()
+            });
+        }
+
+        config
+    }
+
+    pub fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        value: RwTableTag,
```

Comment:
maybe use `value: &S`, where `S: AsBits<N>` so that this chip can be reused for other tag types?

Reply:
This was an oversight. It should just have been `value: &T`.